File: src/Updraft/UpdraftSecurityGroup.cs
using System.Collections.Generic;
using System.Linq;
using Amazon.EC2;
using Amazon.EC2.Model;
using NLog;

namespace Updraft
{
    /// <summary>
    /// Wraps an AWS security group and provides some helper methods to make
    /// it easier to remove and add ip permissions to the security group.
    /// </summary>
    public class UpdraftSecurityGroup
    {
        private readonly AmazonEC2Client client;
        private readonly string securityGroupName;
        private readonly string vpcId;
        private readonly Logger logger = LogManager.GetCurrentClassLogger();

        public UpdraftSecurityGroup(AmazonEC2Client client, string securityGroupName, string vpcId)
        {
            this.client = client;
            this.vpcId = !string.IsNullOrWhiteSpace(vpcId) ? vpcId : null;
            this.securityGroupName = securityGroupName;
        }

        public void AddIngressPermissions(IEnumerable<IpPermission> permissions)
        {
            foreach (var permission in permissions)
            {
                AddIngressPermission(permission);
            }
        }

        public void RemoveIngressPermissions(IEnumerable<IpPermission> permissions)
        {
            foreach (var permission in permissions)
            {
                RemoveIngressPermission(permission);
            }
        }

        public void AddIngressPermission(IpPermission permission)
        {
            var request = new AuthorizeSecurityGroupIngressRequest
            {
                GroupId = GetSecurityGroup().GroupId,
                IpPermissions = { permission }
            };

            try
            {
                client.AuthorizeSecurityGroupIngress(request);
            }
            catch (AmazonEC2Exception e)
            {
                // If the ingress rule already exists then don't complain. The
                // user might have created it manually or they deleted their
                // .last-config or .last-ip files.
                if (e.ErrorCode.Equals("InvalidPermission.Duplicate"))
                {
                    logger.Warn("Tried to create a duplicate ingress rule for group " + request.GroupId);
                    return;
                }

                throw;
            }
        }

        public void RemoveIngressPermission(IpPermission permission)
        {
            var request = new RevokeSecurityGroupIngressRequest
            {
                GroupId = GetSecurityGroup().GroupId,
                IpPermissions = { permission }
            };

            try
            {
                client.RevokeSecurityGroupIngress(request);
            }
            catch(AmazonEC2Exception e)
            {
                // If the ingress rule didn't exist then don't complain.
                if (e.ErrorCode.Equals("InvalidGroup.NotFound"))
                {
                    logger.Warn("Tried to delete an ingress rule that did not exist for group " + request.GroupId);
                    return;
                }

                throw;
            }
        }

        /// <summary>
        /// Attempts to retrieve the security group or create it if it doesn't
        /// exist.
        /// </summary>
        public void CreateSecurityGroupIfNotExists()
        {
            if (GetSecurityGroup() == null)
            {
                CreateSecurityGroup();
            }
        }

        private SecurityGroup GetSecurityGroup()
        {
            // Attempt to retrive the security group
            var request = new DescribeSecurityGroupsRequest
            {
                Filters = {
                    new Filter
                    {
                        Name = "group-name",
                        Values = { securityGroupName }
                    },

                    new Filter
                    {
                        Name = "vpc-id",
                        Values = { vpcId }
                    }
                }
            };

            return client
                .DescribeSecurityGroups(request)
                .SecurityGroups
                .SingleOrDefault();
        }

        private SecurityGroup CreateSecurityGroup()
        {
            var request = new CreateSecurityGroupRequest(securityGroupName, "Maintained by Updraft")
            {
                VpcId = vpcId
            };

            client.CreateSecurityGroup(request);

            return GetSecurityGroup();
        }
    }
}

File: src/UpdraftFS/UpdraftSecurityGroup.fs
module UpdraftSecurityGroup

open Amazon.EC2
open Amazon.EC2.Model
open System

// todo: logging
type UpdraftSecurityGroup =
    { client: AmazonEC2Client
      securityGroupName: string
      vpcId: string }

let create client securityGroupName vpcId =
    { client = client
      securityGroupName = securityGroupName
      vpcId = vpcId }

let getSecurityGroup (updraftSecurityGroup: UpdraftSecurityGroup) =
    let request = new DescribeSecurityGroupsRequest ()
    request.Filters.Add(new Filter (Name = "group-name", Values = new System.Collections.Generic.List<string>([updraftSecurityGroup.securityGroupName])))
    request.Filters.Add(new Filter (Name = "vpc-id", Values = new System.Collections.Generic.List<string>([updraftSecurityGroup.vpcId])))
    updraftSecurityGroup.client
        .DescribeSecurityGroups(request)
        .SecurityGroups
        |> Seq.tryHead

let addIngressPermission (updraftSecurityGroup: UpdraftSecurityGroup) permission =
    match getSecurityGroup updraftSecurityGroup with
    | Some securityGroup ->
        let request = new AuthorizeSecurityGroupIngressRequest ( GroupId = securityGroup.GroupId )
        request.IpPermissions.Add(permission)
        try
            updraftSecurityGroup.client.AuthorizeSecurityGroupIngress(request) |> ignore
        with
        // If the ingress rule already exists then don't complain. The
        // user might have created it manually or they deleted their
        // .last-config or .last-ip files.
        | :? AmazonEC2Exception as e when e.ErrorCode.Equals("InvalidPermission.Duplicate") -> () // todo: logging
    | None -> raise (ApplicationException "Can't get security group to which I want to add the ingress permission")

let addIngressPermissions updraftSecurityGroup permissions =
    List.iter (addIngressPermission updraftSecurityGroup) permissions

let removeIngressPermission updraftSecurityGroup permission =
    match getSecurityGroup updraftSecurityGroup with
    | Some securityGroup ->
        let request = new RevokeSecurityGroupIngressRequest ( GroupId = securityGroup.GroupId )
        request.IpPermissions.Add(permission)
        try
            updraftSecurityGroup.client.RevokeSecurityGroupIngress(request) |> ignore
        with
        // If the ingress rule didn't exist then don't complain.
        | :? AmazonEC2Exception as e when e.ErrorCode.Equals("InvalidPermission.NotFound") -> () // todo: logging
        // If the group which should contain this rule does not
        // exist then don't complain.
        | :? AmazonEC2Exception as e when e.ErrorCode.Equals("InvalidGroup.NotFound") -> () // todo: logging
    | None -> raise (ApplicationException "Can't get security group from which I want to remove the ingress permission")

let removeIngressPermissions updraftSecurityGroup permissions =
    List.iter (removeIngressPermission updraftSecurityGroup) permissions

let createSecurityGroup (updraftSecurityGroup: UpdraftSecurityGroup) =
    let request = new CreateSecurityGroupRequest(updraftSecurityGroup.securityGroupName, "Maintained by Updraft", VpcId = updraftSecurityGroup.vpcId)
    updraftSecurityGroup.client.CreateSecurityGroup request |> ignore

let createSecurityGroupIfNotExists updraftSecurityGroup = 
    match getSecurityGroup updraftSecurityGroup with
    | Some _ -> ()
    | None -> createSecurityGroup updraftSecurityGroup
